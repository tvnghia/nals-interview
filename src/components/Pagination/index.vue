<template>
  <ul class="pagination">
    <li
      v-for="page in pages"
      :key="page.index"
      @click="clickPagination(page.page)"
      :class="['pagination__item', { active: page.active }]"
    >
      <a class="pagination__link">{{ page.page }}</a>
    </li>
  </ul>
</template>

<script>
const PAGE_TOTAL_DUMMY = 5

export default {
  name: 'Pagination',

  props: {
    value: {
      type: Number,
      require: true
    }
  },

  computed: {
    pages() {
      const pages = []
      for (let i = 0; i < PAGE_TOTAL_DUMMY; i++) {
        pages.push({
          active: this.currentPage === i + 1,
          page: i + 1
        })
      }
      return pages
    },

    currentPage() {
      return parseInt(this.$route.query.page)
    }
  },

  methods: {
    clickPagination(page) {
      if (page === this.currentPage) return

      this.$emit('input', page)
    }
  }
}
</script>
