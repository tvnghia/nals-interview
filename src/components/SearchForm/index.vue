<template>
  <input class="form-control u-w-300 u-mr-16" type="text" v-model="valueSearch" />
</template>

<script>
import debounceSearch from '@/mixins/debounce-search'

const TIME_DEBOUNCE = 1000

export default {
  name: 'SearchForm',

  mixins: [debounceSearch],

  data() {
    return {
      valueSearch: this.$route.query.search || ''
    }
  },

  watch: {
    valueSearch: {
      handler(value) {
        // redirect to page 1 when searching
        this.handleDebounceSearch(
          () => this.$emit('on-search-blogs', { search: encodeURI(value), page: 1 }),
          value,
          TIME_DEBOUNCE
        )
      }
    }
  }
}
</script>

<style lang="scss" scoped>
.u-w-300 {
  width: 300px;
}
</style>
