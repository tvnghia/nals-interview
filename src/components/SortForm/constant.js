export const SORT_RADIOS = Object.freeze([
  {
    label: 'Ascending',
    value: 'asc'
  },

  {
    label: 'Descending',
    value: 'desc'
  },

  {
    label: 'None',
    value: 'none'
  }
])
