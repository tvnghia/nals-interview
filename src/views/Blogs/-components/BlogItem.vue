<template>
  <div class="media u-mt-16">
    <img class="media__image" :src="blog.image" alt="blog image" />

    <div class="media__body">
      <p class="headline--md">{{ blog.title }}</p>
      <p class="u-mt-8">{{ blog.content }}</p>
      <p class="headline--sm u-mt-8">Created at: {{ $filters.moment_yyyy_mm_dd(blog.createdAt) }}</p>
    </div>
  </div>
</template>

<script>
export default {
  name: 'BlogItem',

  props: {
    blog: {
      type: Object,
      require: true
    }
  }
}
</script>
