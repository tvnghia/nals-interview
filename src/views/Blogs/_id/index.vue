<template>
  <div class="u-mt-32">
    <h2 class="headline--lg">Blog detail</h2>

    <blog-item v-show="loaded" :blog="blog" />

    <img v-show="!loaded" width="50" class="u-mt-16" src="@/assets/images/loading-buffering.gif" alt="loading gif" />
  </div>
</template>

<script>
import { mapActions, mapState } from 'vuex'
const BlogItem = () => import('../-components/BlogItem')

export default {
  name: 'Blog',

  components: {
    BlogItem
  },

  data() {
    return {
      loaded: false
    }
  },

  computed: {
    ...mapState('blogs', ['blog'])
  },

  methods: {
    ...mapActions('blogs', ['getBlogDetail'])
  },

  created() {
    this.getBlogDetail(this.$route.params.id).finally(() => (this.loaded = true))
  }
}
</script>
