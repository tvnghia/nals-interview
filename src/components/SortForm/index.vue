<template>
  <div class="sort-form u-mt-16">
    <p class="sort-form__title">Sort blogs:</p>

    <div v-for="option in sortOption" :key="option.index" class="u-flex u-items-center">
      <input type="radio" name="sort_blogs" :id="option.id" :value="option.value" v-model="picked" />

      <label :for="option.id">{{ option.label }}</label>
    </div>
  </div>
</template>

<script>
import { SORT_RADIOS } from './constant'

export default {
  name: 'SortForm',

  props: {
    value: {
      type: String,
      default: 'asc'
    }
  },

  data() {
    return {
      sortOption: SORT_RADIOS,
      picked: this.value
    }
  },

  watch: {
    picked: {
      handler(value) {
        this.$emit('input', value)
      }
    }
  }
}
</script>

<style lang="scss" scoped>
.sort-form {
  font-size: 14px;

  &__title {
    font-size: 18px;
    font-weight: 500;
  }

  input {
    margin-right: 4px;
  }

  &,
  input,
  label {
    cursor: pointer;
  }
}
</style>
