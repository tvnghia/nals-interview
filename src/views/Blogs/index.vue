<template>
  <div class="u-mt-32">
    <h1 class="headline--lg">My blogs</h1>

    <div class="u-flex u-items-center u-mt-16">
      <search-form @on-search-blogs="onSearchBlogs" />
      <pagination @on-click-pagination="onClickPagination" />
    </div>

    <sort-form @on-sort-blogs="onSortBlogs" />

    <ul v-if="blogs.length > 0" class="u-pt-40">
      <router-link 
        v-for="item in blogs" 
        :key="item.id" 
        class="media"
        tag="li"
        :to="`/blogs/${item.id}`"
        exact
      >
        <img class="media__image" :src="item.image" alt="image" />
        <div class="media__body">
          <p class="headline--md">{{ item.title }}</p>
          <p class="u-mt-8">{{ item.content }}</p>
          <p class="headline--sm u-mt-8">Created at: {{ $filters.moment_yyyy_mm_dd(item.createdAt) }}</p>
        </div>
      </router-link>
    </ul>

    <p v-else class="u-mt-16 headline--md">Have no blogs</p>
  </div>
</template>
<script>
import { mapState, mapActions } from 'vuex'
import Pagination from '@/components/Pagination'
import SearchForm from '@/components/SearchForm'
import SortForm from '@/components/SortForm'

import { deepCopy } from '@/helpers/json-parser'

export default {
  name: 'Blogs',

  components: {
    Pagination,
    SearchForm,
    SortForm
  },

  data() {
    return {
      paramsRequest: {
        page: this.$route.query.page || 1,
        search: this.$route.query.search || '',
        sortBy: this.$route.query.sortBy === 'none' ? '' : this.$route.query.sortBy,
        order: this.$route.query.order || ''
      }
    }
  },

  computed: {
    ...mapState('blogs', ['blogs'])
  },

  methods: {
    ...mapActions('blogs', ['getBlogs']),

    updateParamsRequest(queryKey) {
      this.$router.push({ query: { ...this.$route.query, ...queryKey } })
      this.paramsRequest = Object.assign(this.paramsRequest, queryKey)
    },

    onClickPagination(page) {
      this.updateParamsRequest({ page })
    },

    onSearchBlogs(searchKey) {
      this.updateParamsRequest({ search: searchKey, page: 1 })
    },

    onSortBlogs(sortKey) {
      const sortBy = sortKey === 'none' ? '' : 'createdAt'
      this.updateParamsRequest({ sortBy, order: sortKey  })
    }
  },

  watch: {
    paramsRequest: {
      handler(value) {
        this.getBlogs(deepCopy(value))
      },

      deep: true
    }
  },

  created() {
    if (!this.$route.query.page) this.$router.push({ query: { ...this.$route.query, page: 1 } })

    this.getBlogs(this.paramsRequest)
  }
}
</script>
