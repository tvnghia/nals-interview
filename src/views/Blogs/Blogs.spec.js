import { shallowMount } from '@vue/test-utils'
import wrapperOps from '@/config-test/wrapper'
import Blogs from '@/views/Blogs'
import Pagination from '@/components/Pagination'
import SearchForm from '@/components/SearchForm'
import SortForm from '@/components/SortForm'
import BlogItem from '@/views/Blogs/-components/BlogItem'

const getBlogs = jest.spyOn(Blogs.methods, 'getBlogs').mockResolvedValue()

describe('Blogs.vue', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(Blogs, wrapperOps)
  })

  afterEach(() => {
    wrapper.vm.$destroy()
  })

  describe('when Blogs was mounted', () => {
    it('should render correctly', () => {
      expect(wrapper.exists()).toBeTruthy()
      expect(wrapper).toMatchSnapshot()
    })

    it('should render child components', () => {
      expect(wrapper.findComponent(Pagination).exists()).toBeTruthy()
      expect(wrapper.findComponent(SearchForm).exists()).toBeTruthy()
      expect(wrapper.findComponent(SortForm).exists()).toBeTruthy()
      expect(wrapper.findComponent(BlogItem).exists()).not.toBeTruthy()
    })

    it('should call getBlogs first time', () => {
      expect(getBlogs).toHaveBeenCalledTimes(1)
    })
  })

  describe('watch paramsRequest', () => {
    it('should getBlogs called', async () => {
      await wrapper.setData({
        paramsRequest: {
          page: 2,
          search: 'book',
          order: 'desc'
        }
      })
      expect(getBlogs).toHaveBeenCalled()
    })
  })
})
