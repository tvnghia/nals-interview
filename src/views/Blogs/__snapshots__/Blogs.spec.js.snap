// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blogs.vue when Blogs was mounted should render correctly 1`] = `
<div class="u-mt-32">
  <h1 class="headline--lg">My blogs</h1>
  <div class="u-flex u-items-center u-mt-16"><input type="text" class="form-control u-w-300 u-mr-16">
    <ul class="pagination">
      <li class="pagination__item active"><a class="pagination__link">1</a></li>
      <li class="pagination__item"><a class="pagination__link">2</a></li>
      <li class="pagination__item"><a class="pagination__link">3</a></li>
      <li class="pagination__item"><a class="pagination__link">4</a></li>
      <li class="pagination__item"><a class="pagination__link">5</a></li>
    </ul>
  </div>
  <div class="sort-form u-mt-16">
    <p class="sort-form__title">Sort blogs:</p>
    <div class="u-flex u-items-center"><input type="radio" name="sort_blogs" id="1" value="asc"> <label for="1">Ascending</label></div>
    <div class="u-flex u-items-center"><input type="radio" name="sort_blogs" id="2" value="desc"> <label for="2">Descending</label></div>
  </div>
  <p class="u-mt-16 headline--md">Have no blog</p>
</div>
`;
